lib/ansible/modules/crypto/get_certificate.py

```diff
@@ -0,0 +1,200 @@
+    description: The algorythm used to sign the cert
+    returned: success
+    type: string
+subject:
+    description: An alias of 'cert'
+    returned: success
+    type: dict
+version:
+    description: The version number of the certificate
+    returned: success
+    type: string
+'''
+
+EXAMPLES = '''
+- name: Get the cert from an RDP port
+  get_certificate:
+    host: "1.2.3.4"
+    port: 3389
+  delegate_to: localhost
+  run_once: true
+  register: cert
+
+- name: Get a cert from an https port
+  get_certificate:
+    host: "www.google.com"
+    port: 443
+  delegate_to: localhost
+  run_once: true
+  register: cert
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+
+from os.path import isfile
+from ssl import get_server_certificate
+from socket import setdefaulttimeout
+
+try:
+    import OpenSSL
+    from OpenSSL import crypto
+except ImportError:
+    pyopenssl_found = False
+else:
+    pyopenssl_found = True
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            ca_certs=dict(required=False, default=None),
+            host=dict(required=True),
+            port=dict(required=True),
+            timeout=dict(required=False, type='int', default=10),
+        ),
+    )
+
+    ca_certs = module.params.get('ca_certs')
+    host = module.params.get('host')
+    port = module.params.get('port')
+    timeout = module.params.get('timeout')
+
+    result = dict(
+        changed=False,
+        subject=''
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg='the python pyOpenSSL module is required')
+
+    if timeout:
+        setdefaulttimeout(timeout)
+
+    if ca_certs:
+        if not isfile(ca_certs):
+            module.fail_json(msg="ca_certs file does not exist")
+
+    try:
+        cert = get_server_certificate((host, port), ca_certs=ca_certs)
+        x509 = crypto.load_certificate(crypto.FILETYPE_PEM, cert)
+    except Exception as e:
+        module.fail_json(msg="Failed to get cert from port with error: {0}".format(e))
+
+    result['cert'] = {}
+    for component in x509.get_subject().get_components():
+        result['cert'][component[0]] = component[1]
+
+    # We would need to pass bytes a digest name to get a digest of the cert
+    # result['digest'] = x509.digest()
+
+    result['expired'] = x509.has_expired()
+
+    result['extensions'] = []
+    extension_count = x509.get_extension_count()
+    for index in range(0, extension_count):
+        extension = x509.get_extension(index)
+        result['extensions'].append({
+            'critical': extension.get_critical(),
+            'asn1_data': extension.get_data(),
+            'name': extension.get_short_name(),
+        })
+
+    result['issuer'] = {}
+    for component in x509.get_issuer().get_components():
+        result['issuer'][component[0]] = component[1]
+
+    result['not_after'] = x509.get_notAfter()
+    result['not_before'] = x509.get_notBefore()
+
+    # result['public_key'] = x509.get_pubkey()
```

Comment:
Why is this commented out?

Reply:
This returns a PKey object which we can't just put into the result dict (you get a stack). I can add a comment like I did for the above message:
    # This retuns a PKey object which we can put directly into the result
    # In the future, if someone asks for more details about the PKey,
    # We could pull specific data out about the cert.

Reply:
I'd rather just remove it...

Reply:
Fixed.